### ISSUE TITLE
byte-code level incompatibility with Java <= 13

### ISSUE THREAD

**Issue by BrunoBonacci** [jgpc42/jmh-clojure#7]
Hi,

The recent update of `jmh-clojure` specifically version `0.3.0` contains pre-compiled classes in the `jmh-clojure-0.3.0.jar` jar file which are compiled with Java 14.
This causes an incompatibility with anything that is running with Java 13 or earlier.

This is the error i get:

```
Syntax error (UnsupportedClassVersionError) compiling at (jmh/util.clj:1:1).
io/github/jgpc42/jmh/DelegateClassLoader has been compiled by a more recent version of the Java Runtime (class file version 58.0), this version of the Java Runtime only recognizes class file versions up to 57.0
```

the `jmh-clojure` jar contains a few precompiled classes

```
jar -tvf ~/.m2/repository/jmh-clojure/jmh-clojure/0.3.0/jmh-clojure-0.3.0.jar
[...]
  1405 Wed Jul 15 11:23:36 BST 2020 io/github/jgpc42/jmh/DelegateClassLoader.class
   927 Wed Jul 15 11:23:36 BST 2020 io/github/jgpc42/jmh/sample/Benchmarks.class
   629 Wed Jul 15 11:23:36 BST 2020 io/github/jgpc42/jmh/sample/Counters.class
  2125 Wed Jul 15 11:23:36 BST 2020 io/github/jgpc42/jmh/Util.class
[...]
```

the version of these classes is 58:
```
$ javap --class-path ~/.m2/repository/jmh-clojure/jmh-clojure/0.3.0/jmh-clojure-0.3.0.jar -v io.github.jgpc42.jmh.Util | head
Classfile jar:file:///Users/bruno/.m2/repository/jmh-clojure/jmh-clojure/0.3.0/jmh-clojure-0.3.0.jar!/io/github/jgpc42/jmh/Util.class
  Last modified 15 Jul 2020; size 2125 bytes
  SHA-256 checksum dc232c06a3c435626351cbcf2db6e4d8746da841463e4e9a3582a44f6c80e208
  Compiled from "Util.java"
public final class io.github.jgpc42.jmh.Util
  minor version: 0
  major version: 58
```

Is this accidental or you just want to drop the compatibility with older Java version?

If it is accidental I suggest adding the desired output version in the `project.clj` as follow:

``` clojure
  :javac-options ["-target" "1.8" "-source" "1.8"]
```

Otherwise please document the compatibility change in your `CHANGELOG.md`

**Comment by jgpc42** [comment]
Thanks for catching this!

I forgot to update my project file after I recently updated my local JDK version from 8 to 14. Obviously, this wasn't caught by the [CI runner](https://github.com/jgpc42/jmh-clojure/blob/master/.github/workflows/test.yml) due to it compiling the Java from source for each test.

I will have `0.3.1` out in a little bit.

**Comment by jgpc42** [comment]
Please try `0.3.1-SNAPSHOT` when you have some time. This is working on my end, I just want to make sure before I release. Thanks again.

**Comment by BrunoBonacci** [comment]
i'm getting this error now

```
Syntax error (ClassNotFoundException) compiling at (insn/util.clj:1:1).
org.objectweb.asm.ConstantDynamic
```

Additionally, [this](https://github.com/jgpc42/lein-jmh/blob/master/resources/version.edn) was causing some problem as it was forcing a particular version of jmh-clojure

**Comment by jgpc42** [comment]
I can't understand why this is happening. `insn 0.4.0` uses `asm 7.0`, which provides the class in question... It seems like there is a version conflict. Do any of your other dependencies depend on `asm`? I recently had a similar problem trying to use [codox](https://github.com/weavejester/codox) with `jmh-clojure` due to it depending on [`asm-all 5.0.3`](https://mvnrepository.com/artifact/org.ow2.asm/asm-all/5.0.3).

Can you provide a repro, or at least show me your `project.clj`?

**Comment by BrunoBonacci** [comment]
you are right, there is `asm-5.2` in the classpath but is not part of any of the dependencies.
Need to figure out where does it get injected. I think you can release, meanwhile I will fix my dependencies.

**Comment by BrunoBonacci** [comment]
I found the source of the conflict (cider/refactor-nrepl transitive dependency via `[org.clojure/tools.analyzer.jvm "0.7.3"]`) and fixed.
I tested the version `0.3.1-SNAPSHOT` and it works!! thx

**Comment by jgpc42** [comment]
I released [`0.3.1`](https://clojars.org/jmh-clojure).

I also released [`lein-jmh 0.2.8`](https://clojars.org/lein-jmh) that uses this new version as well. I added its own [CI runner](https://github.com/jgpc42/lein-jmh/actions/runs/174809114) that will catch any bugs like this in the future.

Thanks again for all your help. Feel free to reopen this if you run into any more problems. 👍

**Comment by BrunoBonacci** [comment]
Great thanks,
one way to avoid classpath collision is to use [MrAnderson](https://github.com/benedekfazekas/mranderson).
It shadows the dependencies in your project and reduces the classpath collision problem.

**Comment by jgpc42** [comment]
I hadn't heard of `mranderson` before. I definitely will consider its use in a future version. Thanks for the info!

**Comment by liuchong** [comment]
I have same [problem](https://github.com/jgpc42/jmh-clojure/issues/7#issuecomment-660670233), Is ` [org.ow2.asm/asm "7.0"]` too new against to the popular libraries:

> https://github.com/clojure/core.async/blob/master/project.clj#L8
> https://github.com/clojure/tools.analyzer.jvm/blob/master/project.clj#L12

**Comment by liuchong** [comment]
Anyway, 0.2.6 seems to work.

**Comment by liuchong** [comment]
Mark same problem as https://github.com/jgpc42/jmh-clojure/issues/7#issuecomment-660670233

**Comment by jgpc42** [comment]
Just a quick update. @BrunoBonacci @liuchong 

I've recently pushed [`jmh-clojure 0.4.0-SNAPSHOT`](https://clojars.org/jmh-clojure/versions/0.4.0-SNAPSHOT) (and [`lein-jmh 0.3.0-SNAPSHOT`](https://clojars.org/lein-jmh/versions/0.3.0-SNAPSHOT)). Aside from some new features and fixes, this version downgrades `insn` to `0.2.1`, which uses Clojure's bundled ASM by default. The recent features and bugfixes of later versions don't impact this library. This will alleviate the dependency conflict for the time being until I decide whether to bundle my own ASM (via [mranderson](https://github.com/benedekfazekas/mranderson)) or something else.

Thanks.
